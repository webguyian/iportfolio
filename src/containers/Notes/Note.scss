.note {
  text-align: left;

  &-body {
    display: flex;
    flex-flow: column wrap;
    height: 100%;
    transform: translateY(-30px);
    transition: transform 0.2s ease-in;

    &--offset {
      transform: translateY(0);
    }
  }

  &-timestamp {
    color: $gray-light;
    font-size: $font-size-s;
    text-align: center;
  }

  &-input {
    @include no-outline;
    width: 100%;
    padding: $gutter;
    border: 0;
    background: none;
    color: $white;
    appearance: none;

    &--text {
      @include type-body;
      height: 645px;
      resize: none;
    }

    &--title {
      @include truncate;
      @include type-display;
      padding-top: 10px;
      padding-bottom: 0;
    }
  }

  &-bottom-bar,
  &-top-bar {
    display: flex;
    z-index: 1;
    position: relative;
    align-items: center;
    justify-content: space-between;
    width: 100%;
    padding: 10px 20px;
    background-color: $gray-dark;
  }

  &-bottom-bar {
    margin-top: auto;
    padding: 0 30px;

    > .ui-btn--with-icon {
      font-size: $font-size-m;

      > .ui-icon {
        color: $yellow;
      }
    }
  }
}
