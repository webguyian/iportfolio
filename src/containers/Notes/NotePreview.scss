.note-preview {
  display: block;
  position: relative;
  margin: 0 25px;
  padding: 10px 0;
  border-top: 1px solid $gray-medium;

  &-title-display,
  &-text-display {
    @include truncate;
  }

  &-title-display {
    font-size: $font-size-default;
  }

  &-body {
    display: flex;
    align-items: flex-start;
  }

  &-date-display {
    display: inline-block;
    min-width: 85px;
    color: $gray;
  }

  &-text-display {
    color: $gray-medium;
  }

  > .ui-btn--delete {
    position: absolute;
    top: 0;
    right: -45px;
    width: 0;
    height: 100%;
    padding: 10px;
    transition: width 0.4s linear;
    background-color: $red;
    color: $white;
  }

  &--swiped {
    > .ui-btn--delete {
      right: -25px;
      width: 80px;
    }
  }

  &--swiping {
    .reminder-input {
      outline: none;
    }

    > .ui-btn--delete {
      width: 115%;
    }
  }
}
