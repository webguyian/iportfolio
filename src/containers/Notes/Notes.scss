.notes-app {
  @include stretch-column;

  > .ui-text--display {
    @include type-title($yellow);
  }

  &-list {
    @include unstyled-list;
    width: 100%;
    max-height: 645px;
    overflow-x: hidden;
    overflow-y: scroll;

    &-item {
      padding: 0;
    }
  }

  &-bottom-bar {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
    margin-top: auto;
    padding: 10px 20px;

    > .ui-text--body {
      flex-grow: 1;
      margin: 0;
      font-size: $font-size-s;
      text-align: center;
    }

    > .ui-btn--with-icon {
      margin-left: auto;
      padding: 2px;
      font-size: $font-size-m;

      > .ui-icon-edit {
        color: $yellow;
      }
    }
  }
}
