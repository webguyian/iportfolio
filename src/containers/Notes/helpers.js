import { compareDesc } from 'date-fns';

export const dateSort = (a, b) => compareDesc(a.date, b.date);
