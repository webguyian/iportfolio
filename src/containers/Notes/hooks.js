import { useEffect, useRef, useState } from 'react';
import { useSwipeable } from 'react-swipeable';

export const useRefFocus = (condition = true, initial = null) => {
  const ref = useRef(initial);

  useEffect(() => {
    const element = ref && ref.current;

    if (condition && element) {
      // Add focus to element
      element.focus();
    }
  }, [condition, ref]);

  return ref;
};

export const useSwipeOffset = (initial = 0) => {
  const [offset, setOffset] = useState(initial);
  const maxOffset = -30;

  const handleSwipe = eventInfo => {
    if (eventInfo.dir !== 'Down') {
      // Exit early if not down swipe
      return false;
    }

    setOffset(eventInfo.deltaY);
  };

  const handleSwipeUp = () => {
    if (!offset) {
      // Exit early with no offset
      return false;
    }

    setOffset(0);
  };

  const handlers = useSwipeable({
    onSwipedUp: handleSwipeUp,
    onSwiping: handleSwipe,
    trackMouse: true
  });

  return [handlers, offset < maxOffset, setOffset];
};
