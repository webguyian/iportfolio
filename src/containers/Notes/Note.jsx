import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import Button from 'components/Button/Button';
import DateTime from 'components/DateTime/DateTime';
import Link from 'components/Link/Link';
import Text from 'components/Text/Text';

import { useRefFocus, useSwipeOffset } from 'containers/Notes/hooks';

const Note = props => {
  const { date, onAdd, onBack, onChange, onDelete, title, text } = props;
  const titleInput = useRefFocus(!title.length);
  const textInput = useRef(null);
  const [handlers, hasOffset] = useSwipeOffset();
  const baseClass = 'note';
  const inputClass = `${baseClass}-input`;
  const textInputClass = `${inputClass}--text`;
  const titleInputClass = `${inputClass}--title`;
  const offsetClass = hasOffset && `${baseClass}-body--offset`;
  const timestampFormat = `MMMM d, yyyy 'at' h:mm a`;

  const handleKeyUp = event => {
    if (event.key === 'Enter') {
      textInput.current.focus();
    }
  };

  return (
    <div className={baseClass}>
      <div className={`${baseClass}-top-bar`}>
        <Link back to="/notes" onClick={onBack} />
      </div>
      <div
        className={classNames(`${baseClass}-body`, offsetClass)}
        {...handlers}
      >
        <Text className={`${baseClass}-timestamp`}>
          <DateTime format={timestampFormat}>{date}</DateTime>
        </Text>
        <input
          className={classNames(inputClass, titleInputClass)}
          onChange={({ target }) => onChange('title', target.value)}
          onKeyUp={handleKeyUp}
          ref={titleInput}
          value={title}
        />
        <textarea
          className={classNames(inputClass, textInputClass)}
          onChange={({ target }) => onChange('text', target.value)}
          ref={textInput}
          value={text}
        />
      </div>
      <div className={`${baseClass}-bottom-bar`}>
        <Button icon="trash" onClick={onDelete}>
          Delete note
        </Button>
        <Button icon="edit" onClick={onAdd}>
          New note
        </Button>
      </div>
    </div>
  );
};

Note.propTypes = {
  date: PropTypes.number.isRequired,
  onAdd: PropTypes.func.isRequired,
  onBack: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired,
  title: PropTypes.string,
  text: PropTypes.string
};

export default Note;
