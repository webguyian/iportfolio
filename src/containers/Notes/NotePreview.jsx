import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import Button from 'components/Button/Button';
import DateTime from 'components/DateTime/DateTime';
import Link from 'components/Link/Link';
import Text from 'components/Text/Text';

import { useSwipeOffset } from 'modules/reminders/hooks';

const NotePreview = props => {
  const { note, onClick, onDelete } = props;
  const [handlers, offset] = useSwipeOffset(onDelete.bind(null, note.id));
  const baseClass = 'note-preview';
  const swipedClass = offset && `${baseClass}--swiped`;
  const swipingClass = offset > 80 && `${baseClass}--swiping`;
  const noTextMessage = 'No additional text';
  const noTitleMessage = 'New Note';

  const handleClick = () => {
    if (offset) {
      return false;
    }

    onClick(note);
  };

  return (
    <div
      className={classNames(baseClass, swipedClass, swipingClass)}
      {...handlers}
    >
      <Link
        to={{ pathname: `/notes/${note.id}`, state: { id: note.id } }}
        onClick={handleClick}
      >
        <Text
          className={`${baseClass}-title-display`}
          element="h2"
          type="display"
        >
          {note.title || noTitleMessage}
        </Text>
        <div className={`${baseClass}-body`}>
          <Text className={`${baseClass}-date-display`}>
            <DateTime format="h:mm a" fromNow>
              {note.date}
            </DateTime>
          </Text>
          <Text className={`${baseClass}-text-display`}>
            {note.text || noTextMessage}
          </Text>
        </div>
      </Link>
      <Button className="ui-btn--delete" onClick={onDelete.bind(null, note.id)}>
        Delete
      </Button>
    </div>
  );
};

NotePreview.propTypes = {
  note: PropTypes.object.isRequired,
  onClick: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired
};

export default NotePreview;
