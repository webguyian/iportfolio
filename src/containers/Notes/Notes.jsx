import React from 'react';
import { Route } from 'react-router-dom';

import Button from 'components/Button/Button';
import CountText from 'components/Text/CountText';
import Text from 'components/Text/Text';
import Note from './Note';
import NotePreview from './NotePreview';

import { useNotes, useRefFocus } from 'modules/notes/hooks';

const Notes = () => {
  const [notes, activeNote, setActiveNote, eventHandlers] = useNotes();
  const addNoteButton = useRefFocus(!notes.length);
  const noteCount = notes.length;
  const baseClass = 'notes-app';

  if (activeNote) {
    return (
      <Route path="/notes/:id">
        <Note
          {...activeNote}
          onAdd={eventHandlers.onAdd}
          onBack={eventHandlers.onBack}
          onChange={eventHandlers.onChange.bind(null, activeNote.id)}
          onDelete={eventHandlers.onDelete.bind(null, activeNote.id)}
        />
      </Route>
    );
  }

  return (
    <div className={baseClass}>
      <Text element="h1" type="display">
        Notes
      </Text>
      {noteCount ? (
        <ul className={`${baseClass}-list`}>
          {notes.map(note => (
            <li key={note.id} className={`${baseClass}-list-item`}>
              <NotePreview
                note={note}
                onClick={setActiveNote}
                onDelete={eventHandlers.deleteNote}
              />
            </li>
          ))}
        </ul>
      ) : null}
      <div className={`${baseClass}-bottom-bar`}>
        <CountText count={noteCount} element="p">
          Notes
        </CountText>
        <Button icon="edit" onClick={eventHandlers.onAdd} ref={addNoteButton}>
          New note
        </Button>
      </div>
    </div>
  );
};

export default Notes;
