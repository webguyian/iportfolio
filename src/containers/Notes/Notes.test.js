import React from 'react';
import { act, create } from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';

import { mockTime } from 'utilities/test';
import * as hooks from 'modules/notes/hooks';

import Notes from './Notes';
import Note from './Note';

describe('<Notes />', () => {
  const useRefFocus = hooks.useRefFocus;
  const useNotes = hooks.useNotes;

  hooks.useRefFocus = jest.fn();
  hooks.useNotes = jest.fn();

  const eventHandlers = {
    deleteNote: jest.fn(),
    onAdd: jest.fn(),
    onBack: jest.fn(),
    onChange: jest.fn(),
    onDelete: jest.fn()
  };

  beforeEach(() => {
    jest.clearAllMocks();
    // Reset to original implementation before each test
    hooks.useRefFocus.mockImplementation(useRefFocus);
    hooks.useNotes.mockImplementation(useNotes);
  });

  it('renders correctly', () => {
    const component = create(
      <MemoryRouter>
        <Notes />
      </MemoryRouter>
    );

    expect(component).toMatchSnapshot();
  });

  it('handles add note', () => {
    const component = create(
      <MemoryRouter>
        <Notes />
      </MemoryRouter>
    );
    const button = component.root.findByProps({ icon: 'edit' });

    act(() => {
      button.props.onClick();
    });

    expect(component).toMatchSnapshot();
  });

  it('handles delete note', () => {
    const component = create(
      <MemoryRouter>
        <Notes />
      </MemoryRouter>
    );
    const button = component.root.findByProps({ icon: 'edit' });

    act(() => {
      button.props.onClick();
    });

    const note = component.root.findByType(Note);

    expect(note.props.onDelete).toBeDefined();

    act(() => {
      note.props.onDelete();
    });
  });

  it('handles delete for multiple notes', () => {
    const anotherTime = Number(new Date('2019-11-11T09:33:00'));
    const notes = [
      {
        id: 'k2uj0rc0',
        date: anotherTime,
        title: 'Another note',
        text: 'This is another note'
      },
      {
        id: 'k17zbp9c',
        date: mockTime,
        title: 'Example note',
        text: 'This is a note'
      }
    ];

    hooks.useNotes.mockReturnValue([notes, notes[1], jest.fn(), eventHandlers]);
    const component = create(
      <MemoryRouter initialEntries={['/notes/k2uj0rc0']} initialIndex={0}>
        <Notes />
      </MemoryRouter>
    );

    const note = component.root.findByType(Note);

    expect(note.props.onDelete).toBeDefined();

    act(() => {
      note.props.onDelete();
    });

    expect(component).toMatchSnapshot();
  });

  it('handles change note', () => {
    const component = create(
      <MemoryRouter>
        <Notes />
      </MemoryRouter>
    );
    const button = component.root.findByProps({ icon: 'edit' });

    act(() => {
      button.props.onClick();
    });

    const note = component.root.findByType(Note);

    expect(note.props.onChange).toBeDefined();

    act(() => {
      note.props.onChange('title', 'Updated note');
    });

    expect(component).toMatchSnapshot();
  });

  it('handles back to notes', () => {
    const notes = [
      {
        id: 'k17zbp9c',
        date: mockTime,
        title: 'Example note',
        text: 'This is a note'
      }
    ];

    hooks.useNotes.mockReturnValue([notes, notes[0], jest.fn(), eventHandlers]);
    const component = create(
      <MemoryRouter initialEntries={['/notes/k17zbp9c']} initialIndex={0}>
        <Notes />
      </MemoryRouter>
    );

    const note = component.root.findByType(Note);

    expect(note.props.onBack).toBeDefined();

    act(() => {
      note.props.onBack();
    });

    expect(component).toMatchSnapshot();
  });
});
