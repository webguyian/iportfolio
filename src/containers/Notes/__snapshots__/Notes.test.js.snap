// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notes /> handles add note 1`] = `
<div
  className="note"
>
  <div
    className="note-top-bar"
  >
    <Link
      back={true}
      onClick={[Function]}
      to="/notes"
    />
  </div>
  <div
    className="note-body"
    onMouseDown={[Function]}
  >
    <span
      className="ui-text ui-text--body note-timestamp"
    >
      October 1, 2019 at 3:11 PM
    </span>
    <input
      className="note-input note-input--title"
      onChange={[Function]}
      onKeyUp={[Function]}
      value=""
    />
    <textarea
      className="note-input note-input--text"
      onChange={[Function]}
      value=""
    />
  </div>
  <div
    className="note-bottom-bar"
  >
    <button
      className="ui-btn ui-btn--with-icon"
      onClick={[Function]}
      type="button"
    >
      <i
        className="ui-icon ui-icon-trash fas fa-trash"
      />
      <span
        className="ui-text ui-text--accessible"
      >
        Delete note
      </span>
    </button>
    <button
      className="ui-btn ui-btn--with-icon"
      onClick={[Function]}
      type="button"
    >
      <i
        className="ui-icon ui-icon-edit fas fa-edit"
      />
      <span
        className="ui-text ui-text--accessible"
      >
        New note
      </span>
    </button>
  </div>
</div>
`;

exports[`<Notes /> handles back to notes 1`] = `
<div
  className="note"
>
  <div
    className="note-top-bar"
  >
    <Link
      back={true}
      onClick={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      to="/notes"
    />
  </div>
  <div
    className="note-body"
    onMouseDown={[Function]}
  >
    <span
      className="ui-text ui-text--body note-timestamp"
    >
      October 1, 2019 at 3:11 PM
    </span>
    <input
      className="note-input note-input--title"
      onChange={[Function]}
      onKeyUp={[Function]}
      value="Example note"
    />
    <textarea
      className="note-input note-input--text"
      onChange={[Function]}
      value="This is a note"
    />
  </div>
  <div
    className="note-bottom-bar"
  >
    <button
      className="ui-btn ui-btn--with-icon"
      onClick={[Function]}
      type="button"
    >
      <i
        className="ui-icon ui-icon-trash fas fa-trash"
      />
      <span
        className="ui-text ui-text--accessible"
      >
        Delete note
      </span>
    </button>
    <button
      className="ui-btn ui-btn--with-icon"
      onClick={[MockFunction]}
      type="button"
    >
      <i
        className="ui-icon ui-icon-edit fas fa-edit"
      />
      <span
        className="ui-text ui-text--accessible"
      >
        New note
      </span>
    </button>
  </div>
</div>
`;

exports[`<Notes /> handles change note 1`] = `
<div
  className="note"
>
  <div
    className="note-top-bar"
  >
    <Link
      back={true}
      onClick={[Function]}
      to="/notes"
    />
  </div>
  <div
    className="note-body"
    onMouseDown={[Function]}
  >
    <span
      className="ui-text ui-text--body note-timestamp"
    >
      October 1, 2019 at 3:11 PM
    </span>
    <input
      className="note-input note-input--title"
      onChange={[Function]}
      onKeyUp={[Function]}
      value="Updated note"
    />
    <textarea
      className="note-input note-input--text"
      onChange={[Function]}
      value=""
    />
  </div>
  <div
    className="note-bottom-bar"
  >
    <button
      className="ui-btn ui-btn--with-icon"
      onClick={[Function]}
      type="button"
    >
      <i
        className="ui-icon ui-icon-trash fas fa-trash"
      />
      <span
        className="ui-text ui-text--accessible"
      >
        Delete note
      </span>
    </button>
    <button
      className="ui-btn ui-btn--with-icon"
      onClick={[Function]}
      type="button"
    >
      <i
        className="ui-icon ui-icon-edit fas fa-edit"
      />
      <span
        className="ui-text ui-text--accessible"
      >
        New note
      </span>
    </button>
  </div>
</div>
`;

exports[`<Notes /> handles delete for multiple notes 1`] = `
<div
  className="note"
>
  <div
    className="note-top-bar"
  >
    <Link
      back={true}
      onClick={[MockFunction]}
      to="/notes"
    />
  </div>
  <div
    className="note-body"
    onMouseDown={[Function]}
  >
    <span
      className="ui-text ui-text--body note-timestamp"
    >
      October 1, 2019 at 3:11 PM
    </span>
    <input
      className="note-input note-input--title"
      onChange={[Function]}
      onKeyUp={[Function]}
      value="Example note"
    />
    <textarea
      className="note-input note-input--text"
      onChange={[Function]}
      value="This is a note"
    />
  </div>
  <div
    className="note-bottom-bar"
  >
    <button
      className="ui-btn ui-btn--with-icon"
      onClick={[Function]}
      type="button"
    >
      <i
        className="ui-icon ui-icon-trash fas fa-trash"
      />
      <span
        className="ui-text ui-text--accessible"
      >
        Delete note
      </span>
    </button>
    <button
      className="ui-btn ui-btn--with-icon"
      onClick={[MockFunction]}
      type="button"
    >
      <i
        className="ui-icon ui-icon-edit fas fa-edit"
      />
      <span
        className="ui-text ui-text--accessible"
      >
        New note
      </span>
    </button>
  </div>
</div>
`;

exports[`<Notes /> renders correctly 1`] = `
<div
  className="notes-app"
>
  <h1
    className="ui-text ui-text--display"
  >
    Notes
  </h1>
  <div
    className="notes-app-bottom-bar"
  >
    <p
      className="ui-text ui-text--body"
    >
      No Notes
    </p>
    <button
      className="ui-btn ui-btn--with-icon"
      onClick={[Function]}
      type="button"
    >
      <i
        className="ui-icon ui-icon-edit fas fa-edit"
      />
      <span
        className="ui-text ui-text--accessible"
      >
        New note
      </span>
    </button>
  </div>
</div>
`;
